### PR TITLE
Devel20240409

### LINKED ISSUE #8975: Dynamic audiovm switching
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)

Currently, changing audiovm requires target qube restart in practice. Technically, it is possible to change it without qube restart, if you know which services to restart in what order etc, but that's very much not obvious. This also includes the case of restarting said audiovm.

### The solution you'd like

Changing audiovm at runtime should be seamless, as is changing netvm.

### The value to a user, and who that user might be

Easier to use non-dom0 audiovm, mostly relevant for USB audio devices. Related to #7750, #8093.

### Details

Implementation-wise, this requires:

- [x] dom0 changing relevant qubesdb (or else) entry when audiovm is changed or (re)started
- [x] pulseaudio agent watching said qubesdb entry and re-establishing vchan connection when needed
- [x] pipewire agent doing the same
- [x] stubdomain (currently also pulseaudio) agent doing the same - a caveat here, currently stubdomain doesn't have qubesdb at all (and has hardcoded domid=0 as audiovm), but it does have qrexec-agent

### PR DESCRIPTION
**Pull request by fepitre**
Fixes https://github.com/QubesOS/qubes-issues/issues/8975

### CODE REVIEW

**Comment by marmarek** [comment]
Audio playback after switching hangs with a lot of messages in stubdomain:
```
pulseaudio: pa_threaded_mainloop_lock failed
pulseaudio: Reason: Entity killed
```
This is a test for switching from sys-audio to dom0. I do see calling `qubes.SetAudioVM+0` in stubdomain log, and there is new module-vchan-sink load, and I think it also connects to pacat in dom0.

File: rootfs/qubes.SetAudioVM [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+set -e
+
+# Default to dom0 if called with empty arg
+domid="${1:-0}"
+
+export PULSE_SERVER=unix:/tmp/pa.sock
+
+pactl unload-module module-vchan-sink
```
Comment (marmarek) [inline rootfs/qubes.SetAudioVM:13]: wouldn't this fail if changing from none to something? the module is not loaded at that time...

Comment (fepitre) [reply rootfs/qubes.SetAudioVM:13]: It looks like unloading module that is not loaded already still work and exit with 0.

Comment (marmarek) [reply rootfs/qubes.SetAudioVM:13]: Looks like pactl exits with 0 if module is already unloaded...

Comment (fepitre) [reply rootfs/qubes.SetAudioVM:13]: For example, in the stubdom shell:
```
pactl unload-module module-vchan-sink
Failed to load cookie file from cookie: No such file or directory
Failed to unload module: Module module-vchan-sink not loaded
echo $?
0
pactl unload-module module-vchan-sink
Failed to load cookie file from cookie: No such file or directory
Failed to unload module: Module module-vchan-sink not loaded
echo $?
0
```

**Comment by marmarek** [comment]
This doesn't work:
```
[2024-05-07 12:28:30] + '[' 80 -ge 0 ]
[2024-05-07 12:28:30] + module_vchan_args='-L module-vchan-sink domid=80'
[2024-05-07 12:28:30] + pulseaudio '--use-pid-file=no' '--daemonize=no' '--exit-idle-time=-1' -n -L 'module-native-protocol-unix auth-anonymous=1 socket=/tmp/pa.sock' -L module-always-sink -L module-always-source -L module-null-sink -L module-null-source -L module-vchan-sink 'domid=80'
...
E: [pulseaudio] main.c: [1;31mToo many arguments.[0m
```

I think it takes domid= as a separate argument instead of an argument to the vchan-sink module.

**Comment by fepitre** [comment]
> This doesn't work:
> 
> ```
> [2024-05-07 12:28:30] + '[' 80 -ge 0 ]
> [2024-05-07 12:28:30] + module_vchan_args='-L module-vchan-sink domid=80'
> [2024-05-07 12:28:30] + pulseaudio '--use-pid-file=no' '--daemonize=no' '--exit-idle-time=-1' -n -L 'module-native-protocol-unix auth-anonymous=1 socket=/tmp/pa.sock' -L module-always-sink -L module-always-source -L module-null-sink -L module-null-source -L module-vchan-sink 'domid=80'
> ...
> E: [pulseaudio] main.c: �[1;31mToo many arguments.�[0m
> ```
> 
> I think it takes domid= as a separate argument instead of an argument to the vchan-sink module.

I don't see where you have this error?

**Comment by marmarek** [comment]
Recording in HVM doesn't work initially. I see the module-vchan-sink is loaded with the domid parameter (here with domid=0, the default setup), and no any significant errors. But it does work after switching, so I guess the default source is wrong initially (but I haven't verified this guess).

**Comment by marmarek** [comment]
Reordering modules did not helped. Recording in HVM still doesn't work until audiovm change

**Comment by marmarek** [comment]
Now it works :)

**Pull request merged**